### PR TITLE
Fix CPU devices parsing and strategy selection in CLI trainer setup

### LINKED ISSUE #393: CPU runs fail with explicit --devices due to list parsing and fixed DDP strategy in CLI
### Summary
`neural_lam/train_model.py` currently parses `--devices` as a list and passes it directly to Lightning, while also hardcoding `strategy="ddp"`.

On CPU this is invalid in common local setups:
- Lightning expects an integer device count for CPU (not a list like `[1]`).
- DDP is not appropriate as a default for single-device CPU runs.

This causes CPU workflows (including local debugging/CI-style runs) to fail before training starts.

### Why this is urgent
- Blocks CPU-only users from running train/eval via CLI with explicit `--devices`.
- Creates avoidable launch failures in non-GPU environments.
- Prevents simple local reproducibility checks when no CUDA device is available.

### Reproduction
From repo root, run (example):
```bash
python -m neural_lam.train_model --dataset dummy --devices 1 --num_epochs 1
```
(or equivalent CLI invocation using CPU)

### Current behavior
Trainer initialization fails due to invalid CPU `devices` type and/or strategy mismatch.

### Expected behavior
- CPU path should accept `--devices` as a single positive integer count.
- Strategy should be selected contextually:
  - `auto` for CPU and single-device runs
  - `ddp` for multi-device CUDA runs

### Related but not duplicate
There is an existing issue around hardcoded multi-GPU assumptions in tests (#192), but this report targets runtime CLI behavior on CPU with explicit `--devices` parsing + strategy selection.

### Proposed fix
Add centralized device/strategy resolution in CLI before Trainer construction:
- Normalize CPU devices to integer
- Validate CPU does not receive multiple explicit device IDs
- Choose `strategy` based on accelerator/device count

### PR DESCRIPTION
**Pull request by kshirajahere**
## Describe your changes

Fix the CLI trainer setup so CPU runs with explicit `--devices` arguments behave correctly and Lightning strategy selection matches the actual runtime context.

This adds a small helper, `resolve_devices_and_strategy(args_devices, device_name)`, in `neural_lam/train_model.py` that:
- normalizes CPU device configuration to the integer form Lightning expects,
- rejects invalid CPU device specifications such as multiple IDs or non-positive values,
- keeps `strategy="auto"` for CPU and single-device runs,
- uses `strategy="ddp"` only for multi-device CUDA runs.

The motivation is to avoid invalid Trainer configuration for CPU execution while preserving the expected behavior for CUDA launches.

This change has no new external dependencies.

## Issue Link

Closes #393

## Type of change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [ ] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [ ] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [ ] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [ ] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

File: tests/test_cli.py [hunk 1]
```diff
@@ -53,6 +53,40 @@ def test_wandb_resume_not_exposed():
...
 
 
+@pytest.mark.parametrize(
+    "args_devices, device_name, expected_devices, expected_strategy",
+    [
+        (["auto"], "cpu", 1, "auto"),
+        (["auto"], "cuda", "auto", "auto"),
+        (["1"], "cpu", 1, "auto"),
+        (["0"], "cuda", [0], "auto"),
+        (["0", "1"], "cuda", [0, 1], "ddp"),
```
Comment (Sharkyii) [inline tests/test_cli.py:63]:
```suggestion
    "args_devices, device_name, expected_devices, expected_strategy",
    [
        (["auto"], "cpu", 1, "auto"),
        (["auto"], "cuda", "auto", "auto"),
        (["1"], "cpu", 1, "auto"),
        (["0"], "cuda", [0], "auto"),
        (["0", "1"], "cuda", [0, 1], "auto"),
```

File: neural_lam/train_model.py [hunk 2]
```diff
@@ -24,6 +24,38 @@
...
+    if device_name == "cpu":
+        if len(parsed_devices) != 1 or parsed_devices[0] <= 0:
+            raise ValueError(
+                "On CPU, --devices must be a single integer > 0 "
+                "(for example: --devices 1)."
+            )
+        return parsed_devices[0], "auto"
+
+    if len(parsed_devices) > 1:
+        return parsed_devices, "ddp"
```
Comment (Sharkyii) [inline neural_lam/train_model.py:54]: I think line 53 and 54 are of no use as `return parsed_devices, "auto"` is already there and lightning can handle this part

**Comment by Sharkyii** [comment]
The resolve_devices_and_strategy function is solid, it correctly normalizes CPU devices to integers, validates inputs and handles strategy selection.

**Comment by kshirajahere** [comment]
@Sharkyii Thanks for the review.

This behavior is intentional in this PR:

- CPU runs are normalized to a single positive integer device count.

- Single-device CUDA uses strategy auto.

- Multi-device CUDA uses explicit ddp.

So the branch returning ddp for multiple CUDA devices is not redundant. It encodes a deliberate policy for deterministic distributed behavior instead of relying on implicit strategy inference.

For the test case with [0, 1] on CUDA, expected strategy is therefore ddp by design.

If maintainers prefer fully inferred strategy selection for multi-GPU as well, I can update the implementation and tests in a follow-up. That would be a design change rather than a bug fix.

**Comment by Sharkyii** [comment]
@kshirajahere anyways work as far as I am concerned output remains same in both cases

**Comment by kshirajahere** [comment]
@sadamov I request you sir to please take out some time and review my PR :)

**Comment by sadamov** [comment]
@kshirajahere I am taking a few days off, better if somenone else can review.

**Comment by sadamov** [comment]
Closing this @kshirajahere, sorry for the long wait. The `strategy="ddp"` → `"auto"` fix already landed via #193, so the headline bug here is gone. The remaining useful piece is the CPU `--devices` validator (rejecting e.g. `--devices 0 1` on CPU). Happy to reopen and reduce scope to just that 5-line check if you want. Thanks for the work!
